src/test/sql/sqlite_testrunner/sqlite_testrunner.cpp

```diff
@@ -51,21 +51,59 @@ class SQLiteTestRunner : public BaseTestWithParam<std::string> {
       std::string table_file = args.at(0);
       std::string table_name = args.at(1);
 
-      DebugAssert(!StorageManager::get().has_table(table_name), "Table already loaded");
+      // Store loaded tables in a map that basically caches the loaded tables. In case the table
+      // needs to be reloaded (e.g., due to modifications), we also store the file path.
+      _test_table_cache.emplace(table_name, TestTable{load_table(table_file, 10), table_file, false});
 
-      _sqlite->create_table_from_tbl(table_file, table_name);
-
-      std::shared_ptr<Table> table = load_table(table_file, 10);
-      StorageManager::get().add_table(table_name, std::move(table));
+      // Prepare table copy which is later used as the master to copy from.
+      _sqlite->create_table_from_tbl(table_file, table_name + _master_table_suffix);
     }
 
     opossum::Topology::use_numa_topology();
     opossum::CurrentScheduler::set(std::make_shared<opossum::NodeQueueScheduler>());
+  }
+
+ protected:
+  void SetUp() override {  // called once before each test
+    // For proper testing, we reset the storage manager before EVERY test.
+    StorageManager::get().reset();
+
+    for (auto const& [table_name, test_table] : _test_table_cache) {
+      /*
+        Opossum:
+          We start off with cached tables (SetUpTestCase) and add them to the resetted
+          storage manager before each test here. In case tables have been modified, they are
+          removed from the cache and we thus need to reload them from the initial tbl file.
+        SQLite:
+          Drop table and copy the whole table from the master table to reset all accessed tables.
+      */
+      if (test_table.is_dirty) {
+        // 1. reload table from tbl file, 2. add table to storage manager, 3. cache table in map
+        auto reloaded_table = load_table(test_table.filename, 10);
+        StorageManager::get().add_table(table_name, reloaded_table);
+        _test_table_cache[table_name].table = reloaded_table;
+        _test_table_cache[table_name].is_dirty = false;
+      } else {
+        StorageManager::get().add_table(table_name, test_table.table);
+      }
+
+      _sqlite->reset_table_from_copy(table_name, table_name + _master_table_suffix);
```

Comment:
I agree with mrks, the sqlite table should only be copied if it was actually dirty.